File: cpp/CodingTest/CodingTest/VaildParentheses.cpp
#include<iostream>
#include<stack>

using namespace std;

class Solution {
private:
    stack<char> st;
public:
    Solution(){}
    ~Solution() {
        for (int i = 0; i < st.size(); i++) st.pop();
    }
    bool isValid(string s) {
        if (s[0] == ')' || s[0] == ']' || s[0] == '}') {
            return false;
        }
        else if (s[s.size() - 1] == '(' || s[s.size() - 1] == '[' || s[s.size() - 1] == '{') {
            return false;
        }
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
                st.push(s[i]);
                continue;
            }
            if (st.empty()) {
                if (s[i] == ')' || s[i] == '}' || s[i] == ']') return false;
            }
            if (s[i] == ')' && !st.empty() && st.top() != '(') {
                return false;
            }
            if (s[i] == '}' && !st.empty() && st.top() != '{') {
                return false;
            }
            if (s[i] == ']' && !st.empty() && st.top() != '[') {
                return false;
            }
            if(!st.empty()) st.pop();
        }
        if (!st.empty()) return false;

        return true;
    }
};

int main() {
    Solution s;
    cout << s.isValid("()");
	return 0;
}

File: rust/code/leet_code/stack/VaildParentheses.rs
struct Solution;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut st : Vec<char> = Vec::new();
        for c in s.chars(){
            if c == '(' || c == '[' || c == '{' {
                st.push(c);
                continue;
            }
            if st.is_empty(){
                if c == ')' || c == '}' || c == ']' { return false; }
            }
            if c == ')' && !st.is_empty() && st[st.len() - 1] != '(' { return false; }
            if c == '}' && !st.is_empty() && st[st.len() - 1] != '{' { return false; }
            if c == ']' && !st.is_empty() && st[st.len() - 1] != '[' { return false; }
            if !st.is_empty(){
                st.pop();
            }
        }
        if !st.is_empty() { return false; }
        true
    }
}

pub fn run(){
    println!("{}" ,Solution::is_valid("(){()}".parse().unwrap()));
}